server/index.js

```diff
@@ -51,6 +51,94 @@ app.get('/api/products/:productId', (req, res, next) => {
     .catch(err => next(err));
 });
 
+app.get('/api/cart', (req, res, next) => {
+  if (req.session.cartId) {
+    const sessionSql = `
+    select "c"."cartItemId",
+           "c"."price",
+           "p"."productId",
+           "p"."image",
+           "p"."name",
+           "p"."shortDescription"
+      from "cartItems" as "c"
+      join "products" as "p" using ("productId")
+     where "c"."cartId" = $1
+    `;
+    const sessionCartId = [req.session.cartId];
+    db.query(sessionSql, sessionCartId)
+      .then(data => res.json(data.rows));
+  } else {
+    const sql = `
+    select *
+      from "cartItems"
+    `;
+    db.query(sql)
+      .then(data => res.json(data.rows));
+  }
+});
+
+app.post('/api/cart/:productId', (req, res, next) => {
+  if (req.params.productId < 1) next(new ClientError('Product ID must be a positive integer', 400));
+  const sql = `
+  select "price"
+    from "products"
+   where "productId" = $1
+  `;
+  const productId = [req.params.productId];
+  db.query(sql, productId)
+    .then(data => {
+      if (!data.rows) {
+        next(new ClientError('Cannot find a product with that ID', 400));
+      } else {
+        if (req.session.cartId) {
+          return {
+            cartId: req.session.cartId,
+            price: data.rows[0].price
+          };
+        } else {
+          const cartSql = `
+          insert into "carts" ("cartId", "createdAt")
+          values (default, default)
+          returning "cartId"
+          `;
+          return db.query(cartSql)
+            .then(result => {
+              return {
+                cartId: result.rows[0].cartId,
+                price: data.rows[0].price
+              };
+            });
+        }
+      }
+    })
+    .then(data => {
+      req.session.cartId = data.cartId;
+      const cartIdSql = `
+      insert into "cartItems" ("cartId", "productId", "price")
+      values ($1, $2, $3)
+      returning "cartItemId"
+      `;
+      const cartInfo = [data.cartId, req.params.productId, data.price];
+      return db.query(cartIdSql, cartInfo);
+    })
+    .then(result => {
+      const cartItemIdSql = `
+      select "c"."cartItemId",
+            "c"."price",
+            "p"."productId",
+            "p"."image",
+            "p"."name",
+            "p"."shortDescription"
+        from "cartItems" as "c"
+        join "products" as "p" using ("productId")
+      where "c"."cartItemId" = $1
+      `;
+      const cartItemId = [result.rows[0].cartItemId];
+      db.query(cartItemIdSql, cartItemId)
```

Comment:
Be sure to `return` this promise chain in case there is a rejection. Then the error will be handled by the final `catch()` callback.